Eip4844 Point Evaluation precompile

mariaKt opened the pull request:
This PR implements the [point evaluation precompile for EIP-4844](https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile). It relies on [this](https://github.com/runtimeverification/blockchain-k-plugin/pull/209) PR, that implements a new hook for `verify_kzg_proof` in the blockchain-k-plugin.

anvacaru commented on kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md:
```diff
@@ -2062,6 +2064,28 @@ Precompiled Contracts
          <callData> DATA </callData>
       requires lengthBytes( DATA ) =/=Int 213
+
+    syntax PrecompiledOp ::= "KZGPOINTEVAL"
+ // ----------------------------------
```
formatting
```suggestion
    syntax PrecompiledOp ::= "KZGPOINTEVAL"
 // ---------------------------------------
```

anvacaru commented on kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md:
```diff
@@ -2062,6 +2064,28 @@ Precompiled Contracts
+       orBool #kzg2vh(substrBytes(DATA, 96, 144)) =/=K substrBytes(DATA, 0, 32)
+       orBool notBool verifyKZGProof(substrBytes(DATA, 96, 144), substrBytes(DATA, 32, 64), substrBytes(DATA, 64, 96), substrBytes(DATA, 144, 192))
+
+    syntax Bytes ::= #kzg2vh ( Bytes ) [symbol(#kzg2vh), function, total]
+ // -------------------------------------------------------------------------------------------------------------
```
formatting
```suggestion
    syntax Bytes ::= #kzg2vh ( Bytes ) [symbol(#kzg2vh), function, total]
 // ---------------------------------------------------------------------
```

anvacaru commented on kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm.md:
```diff
@@ -2414,6 +2438,7 @@ The intrinsic gas calculation mirrors the style of the YellowPaper (appendix H).
     rule <k> #gasExec(SCHED, ECMUL)     => Gecmul < SCHED > ... </k>
     rule <k> #gasExec(SCHED, ECPAIRING) => Gecpairconst < SCHED > +Int (lengthBytes(DATA) /Int 192) *Int Gecpaircoeff < SCHED > ... </k> <callData> DATA </callData>
     rule <k> #gasExec(SCHED, BLAKE2F)   => Gfround < SCHED > *Int #asWord(#range(DATA, 0, 4) ) ... </k> <callData> DATA </callData>
+    rule <k> #gasExec(_, KZGPOINTEVAL)  => 50000 ... </k>
```
I think it would be better to use a schedule variable here,
```suggestion
    rule <k> #gasExec(SCHED , KZGPOINTEVAL)  => Gpointeval < SCHED > ... </k>
```
Where `Gpointeval`  would be defined in `schedule.md` as
```
rule Gpointeval < DEFAULT > => 0
rule Gpointeval < CANCUN > => 50000
```

This ensures that the `50000` value is preserved and can be updated dynamically in future hardforks without modifying this rule directly.

mariaKt replied:
Got it, I believe I addressed it.

anvacaru replied:
Looks great! Thank you!

dwightguth reviewed (commented):
My only suggestion is to put BLS_MODULUS into a macro.

mariaKt commented:
@anvacaru since your review and addressing your comments, I have made a few small changes.
- added a couple of side conditions to the rules, regarding conditions that the inputs need to satisfy.
- changed a large constant that was used often to a macro.

Robertorosmaninho commented on tests/failing.llvm:
```diff
@@ -37,11 +37,33 @@ BlockchainTests/GeneralStateTests/Pyspecs/cancun/eip4844_blobs/blob_tx_attribute
 BlockchainTests/GeneralStateTests/Pyspecs/cancun/eip4844_blobs/correct_excess_blob_gas_calculation.json,*
 BlockchainTests/GeneralStateTests/Pyspecs/cancun/eip4844_blobs/correct_increasing_blob_gas_costs.json,*
-BlockchainTests/GeneralStateTests/Pyspecs/cancun/eip4844_blobs/external_vectors.json,*
+BlockchainTests/GeneralStateTests/Pyspecs/cancun/eip4844_blobs/external_vectors.json,src/GeneralStateTestsFiller/Pyspecs/cancun/eip4844_blobs/test_point_evaluation_precompile.py::test_external_vectors[fork_Cancun-blockchain_test-versioned_hash_None-verify_kzg_proof_case_correct_proof_1ce8e4f69d5df899]
```
Hey, @mariaKt! Have you investigated why some `external_vectors.json` tests work and others don't?

mariaKt replied:
Hey @Robertorosmaninho . I now have a solution for this (the plugin should be updated soon), and have updated the failing.llvm to remove these tests. Thanks for identifying the tests!

Robertorosmaninho replied:
Thank you for working on this, @mariaKt! Glad that you have solved this!

Robertorosmaninho reviewed (approved):
As soon as you update the `tests/failing.llvm` file, all tests should pass as expected!
